Add async_search.submit to HLRC

cbuescher opened the pull request:
This commit adds a new AsyncSearchClient to the High Level Rest Client which
initially supporst the submitAsyncSearch in its blocking and non-blocking
flavour. Also adding client side request and response objects and parsing code
to parse the xContent output of the client side AsyncSearchResponse together
with parsing roundtrip tests and a simple roundtrip integration test.

Relates to #49091

Commit:
Add async_search.submit to HLRC

This commit adds a new AsyncSearchClient to the High Level Rest Client which
initially supporst the submitAsyncSearch in its blocking and non-blocking
flavour. Also adding client side request and response objects and parsing code
to parse the xContent output of the client side AsyncSearchResponse together
with parsing roundtrip tests and a simple roundtrip integration test.

Relates to #49091

elasticmachine commented:
Pinging @elastic/es-search (:Search/Search)

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Java High Level REST Client)

cbuescher commented on client/rest-high-level/src/test/java/org/elasticsearch/client/asyncsearch/AsyncSearchIT.java:
```diff
@@ -0,0 +1,68 @@
+
+        SearchRequest searchRequest = new SearchRequest(index);
+        // TODO either check that in client request validation already or set it automatically?
+        searchRequest.setCcsMinimizeRoundtrips(false);
```
We should probably not force users to remember to set this to `false`, although `true` seems to be the default. I wonder if we shold either 
* set the value to `false` automatically for every SearchRequest once it is set for an AsyncSearchRequest
* ignore this parameter as request parameter when sending the request
* ignore this paramter on the receiving Rest action regardless of how it is set in the client (maybe warn)

jimczi replied:
Can we copy the original `SubmitAsyncSearchRequest` that creates the `SearchRequest` internally and sets the default appropriately. We can then add a validation client side if needed or let the server-side validation throw error if users override the ccs option ?

cbuescher replied:
Thanks, I didn't see we implicitely set the defaulfs in the server side `org.elasticsearch.xpack.core.search.action.SubmitAsyncSearchRequest` we should do that on the client side as well then. Also see we're overwriting these properties:
```
request.setPreFilterShardSize(1);
request.setBatchedReduceSize(5);
request.requestCache(true);
```
I assume we also want to do that then on the client side as well?

jimczi replied:
yes

jimczi replied:
This can be removed now that we set these values automatically in the ctr ?

Commit:
iter

jimczi commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/SubmitAsyncSearchRequest.java:
```diff
@@ -0,0 +1,87 @@
+
+import java.util.Objects;
+
+public class SubmitAsyncSearchRequest extends TimedRequest {
```
Is it required to extend `TimedRequest` ? What's the advantage of doing so ?

cbuescher replied:
It gives "timeout" and "master_timeout" parameters, but I guess on the submit request they don't make much sense. Removing.

jimczi replied:
+1

jimczi commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/SubmitAsyncSearchRequest.java:
```diff
@@ -0,0 +1,87 @@
+    public SubmitAsyncSearchRequest(SearchRequest searchRequest) {
+        this.searchRequest = searchRequest;
+}
+
```
Can you add javadocs for the options ?

jimczi commented on server/src/main/java/org/elasticsearch/action/search/SearchResponse.java:
```diff
@@ -255,8 +255,12 @@ public XContentBuilder innerToXContent(XContentBuilder builder, Params params) t
 
     public static SearchResponse fromXContent(XContentParser parser) throws IOException {
-        ensureExpectedToken(Token.START_OBJECT, parser.nextToken(), parser::getTokenLocation);
-        parser.nextToken();
```
Can we try to avoid this change ? Maybe using a custom parser in the root object parser for the `response` field ?

cbuescher replied:
I did that at first but found the change in the server object is cleaner. Happy to change to the other solution though, but I think its working around something that we can fix better here.

jimczi replied:
We can also split into `fromXContent` and `fromXContentFragment`, the former would consume the start object and delegate to `fromXContentFragment` ?

cbuescher replied:
We already have `innerFromXContent` which in an earlier version of this PR I exposed and used from the client side, but then we have to explain that there. I'll change the PR back to that version so you can see the options.

jimczi reviewed (commented):
The change looks good to me. I left some comments but @javanna may want to take a look too.

Commit:
Change supertype of SubmitAsyncSearchRequest

Commit:
Adding javadoc

Commit:
Overwrite certain SearchRequest paramters on submit request

Commit:
Change SearchResponse parsing

jimczi commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/SubmitAsyncSearchResponse.java:
```diff
@@ -0,0 +1,222 @@
+/**
+ * A response of an async search request.
+ */
+public class SubmitAsyncSearchResponse implements ToXContentObject  {
```
The response is shared with the `get` api so can you rename like the original `AsyncSearchResponse` ?

cbuescher replied:
+1. Already did that in a follow up PR but I will also do it here already then.

jimczi commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/SubmitAsyncSearchRequest.java:
```diff
@@ -0,0 +1,122 @@
+     * the input {@link SearchRequest}.
+     * @param searchRequest the actual search request to submit
+     */
+    public SubmitAsyncSearchRequest(SearchRequest searchRequest) {
```
The original request takes the indices to search in the constructor rather than the `SearchRequest` that is created internally. Can you adapt it to match the behavior of the transport request ?

cbuescher replied:
I thought about it but decided against it. The transport requests signature ` public SubmitAsyncSearchRequest(SearchSourceBuilder source, String... indices)` doesn't allow setting things in the SearchRequest thats constructed internally, like e.g. `search_type` or `routing`. 
You probably would need to do sth. like this on the client side which I find cumbersome:
```
SearchSourceBuilder sourceBuilder = new SearchSourceBuilder()....;        
SubmitAsyncSearchRequest submitRequest = new SubmitAsyncSearchRequest(indices, sourceBuilder);
submitRequest.getSearchRequest().routing(). 
```
I think on the client side its easier for the user to pass in the full `SearchRequest` rather than having to do things like this. Wdyt?

jimczi replied:
Then maybe we should expose the search request options that we allow to modify directly in the `SubmitAsyncSearchRequest` and hide the internal `searchRequest` ? Whatever we do here I'd prefer to have the same signature between hlrc and transport since they should be equivalent.

cbuescher replied:
> here I'd prefer to have the same signature between hlrc and transport since they should be equivalent.

I'm open to adding a bunch of setters to the client side SubmitAsyncSearchRequest but I wonder if its really necessary. Why do the transport and the client side request ctors need to be equivalent? With the transport client gone I only see the transport request used internally, everything else would go through the RestSubmitAsyncSearchAction which handles creating the transport request, or am I missing something? Maybe @javanna has some ideas here as well.

jimczi replied:
I agree but in an ideal world we'd have a single class here so I prefer if we don't diverge in the hlrc. Exposing the search request directly is tricky since some of the options are not allowed (scroll, ccs) and we change the default internally. Hiding the search request entirely behind direct setters and getters seems safer and would avoid the confusion when building an async search request ?

jimczi commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/SubmitAsyncSearchResponse.java:
```diff
@@ -0,0 +1,222 @@
+        // we should be before the opening START_OBJECT of the response
+        ensureExpectedToken(Token.START_OBJECT, p.currentToken(), p::getTokenLocation);
+        p.nextToken();
+        return SearchResponse.innerFromXContent(p);
```
++

Commit:
Merge branch 'master' into async-search-hlrc

Commit:
Rename response class

Commit:
Don't set CcsMinimizeRountrips in test

Commit:
Use setCleanOnCompletion(false) in AsyncSearchIT to always get an ID back

cbuescher commented:
@jimczi thanks for the review, I pushed some updates and left a reply regarding changing the client-side SubmitAsyncSearchRequest constructor, let me know what you think.

Commit:
Change ctor of SubmitAsyncSearchRequest

cbuescher commented:
I pushed a change with a different ctor, I'm still not a fan of this but added it as a base for discussion.

jimczi commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/SubmitAsyncSearchRequest.java:
```diff
@@ -0,0 +1,169 @@
+    public void allowPartialSearchResults(boolean allowPartialSearchResults) {
+        this.searchRequest.allowPartialSearchResults(allowPartialSearchResults);
+    }
+
```
We should also expose batchedReduceSize since that's the option to use to change the granularity of the progress ?

cbuescher replied:
Sure, I didn't add that because we implicitely set it to 5 in the ctor now, so thought its something we don't want the user to set.
What about PreFilterShardSize and requestCache then, those I also didn't add since we set them to fixed values internally?

jimczi replied:
I would expose batchedReduceSize and requestCache only, preFilterShardSize is not necessary imo (we want to always run the can_match phase in async search).

jimczi reviewed (commented):
I like the new ctr better, I also find it cleaner to not expose the raw search request. Can you explain why you're not happy with it ?

Commit:
Merge branch 'master' into async-search-hlrc

Commit:
Add request validation and tests

Commit:
iter on failing test

Commit:
Expose batchedReduceSize and batchedReduceSize on submit request

cbuescher commented:
@elasticmachine run elasticsearch-ci/1

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/AsyncSearchClient.java:
```diff
@@ -0,0 +1,68 @@
+     * @param listener the listener to be notified upon request completion
+     * @return cancellable that may be used to cancel the request
+     */
+    public Cancellable submitAsyncSearchAsync(SubmitAsyncSearchRequest request, RequestOptions options,
```
I wonder if we need to expose this async async method. Is there a reason why people would call an async API asynchronously?

cbuescher replied:
Please correct me, but as far as I understand it, this is not about the asynchronism of the search request, but the submit request itself. While the _async_search request controlls when and whether it returns by waitForCompletion / keepAlive, this method allows using a listener for wait for the response instead of blocking. So say, a user sends and "_async_search" with waitForCompletion=10s because thats how long its okay for the search to initially take with the first responses, but wants the client call to return immediately nevertheless? I think having this method also makes sense in terms of consistency of the API with other client methods.

javanna replied:
ok lets leave it.

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java:
```diff
@@ -403,13 +403,14 @@ static Request search(SearchRequest searchRequest, String searchEndpoint) throws
         params.withIndicesOptions(searchRequest.indicesOptions());
         params.putParam("search_type", searchRequest.searchType().name().toLowerCase(Locale.ROOT));
         params.putParam("ccs_minimize_roundtrips", Boolean.toString(searchRequest.isCcsMinimizeRoundtrips()));
+        params.putParam("pre_filter_shard_size", Integer.toString(searchRequest.getPreFilterShardSize()));
```
good catch, looks like search was not supporting this parameter.

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/AsyncSearchResponse.java:
```diff
@@ -0,0 +1,222 @@
+    private String id;
+    private final int version;
+    private SearchResponse searchResponse;
+    private ElasticsearchException error;
```
maybe this should be nullable too?

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/SubmitAsyncSearchRequest.java:
```diff
@@ -0,0 +1,211 @@
+    /**
+     * Get the {@link SearchRequest} that this submit request wraps
+     */
+    public SearchRequest getSearchRequest() {
```
I see that whether or not to expose the inner search request was discussed in previous reviews. I am confused though on what direction was chosen. I see that the getters and setters from the search request are copied to the async request, but the inner search request is still exposed through this getter and can be modified directly. Didn't we want to rather hide it from users?

jimczi replied:
+1 to hide

cbuescher replied:
Okay, I will look into that. Will make testing a bit more awkward though because I can't reuse some existing infra then.

jimczi replied:
package protected ? ;)

cbuescher replied:
Work work but then I need to move some classes around. I'd prefer that to adding all those getters to the submit request.

javanna replied:
I think that if we can do without the getter, even better. The changes you had to make to remove it don't look so bad. The problem with keeping it package protected is that theoretically some stuff can still be set directly to the search request, then we need to validate parameters that are not even supported.

cbuescher replied:
This is proving to be tricky. I cannot easily move the AsyncSearchRequestConverters out of the `org.elasticsearch.client` package since they need package private infra from RequestConverters. I can probably move the `SubmitAsyncSearchRequest` into that package though, will take a look what that would change.

cbuescher replied:
I pushed f686a50 which moves the new request and response classes into `org.elasticsearch.client` to be able to use a package protected `getSearchRequest()` to avoid all the boilerplate getters on the new request. I'm unsure what I like best, take a look and let me know which direction you are leaning.

javanna replied:
I  think no getter is  safer. We also do have extra protection on the server side for unsupported values but I think the client should never allow to send unsupported values.

jimczi replied:
I agree with Luca, we should avoid the getter on the search request. However, we should have a getter for all options that we expose so I don't see why you want to avoid them ? If we have a setter, we need to provide a way to access the value.

cbuescher replied:
Ok, I reverted that change back to the one where I removed `getSearchRequest` and added a bunch of getters.

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/SubmitAsyncSearchRequest.java:
```diff
@@ -0,0 +1,211 @@
+    /**
+     * Set the routing values to control the shards that the search will be executed on.
+     */
+    public void routing(String... routings) {
```
annoying nit: I see that some getters and setters use the get and set prefix in this class, while others don't. I get why, but shall we unify the style at least throughout this one class?

javanna commented on client/rest-high-level/src/test/java/org/elasticsearch/client/asyncsearch/AsyncSearchIT.java:
```diff
@@ -0,0 +1,53 @@
+        SearchSourceBuilder sourceBuilder = new SearchSourceBuilder().query(QueryBuilders.matchAllQuery());
+        SubmitAsyncSearchRequest request = new SubmitAsyncSearchRequest(sourceBuilder, index);
+        // 2 sec should be enough to make sure we always complete right away
+        request.setWaitForCompletion(new TimeValue(2, TimeUnit.SECONDS));
```
this should work but it's also risky. What other options we have? 
1) increase the timeout even more to really make sure it's enough, though what is enough
2) accept that we may get a response while the search is still running, in which case is_partial and is_final will have a different value? Maybe don't set cleanOnCompletion and assert that when the id is returned the search is running, and when the id is not returned, both flags are false.

I am not sure either way, maybe  I would go for option 1.

jimczi replied:
We can also have a loop here when get is implemented that submits the initial request and use get until `isRunning` is false on the response ?

cbuescher replied:
I'll go with 1.) then, staying on the save side. I will keep `cleanOnCompletion` though, currently thats the only way I can see to actually get an Id and I'd like to tests its deserialization etc... here if possible (will also need ids in a follow-up PR around get-API)

cbuescher replied:
sorry, missed the previous comment. Loop sounds okay as well, getting a response thats not running any more isn't actually the problem, the oposite is tricky (getting a response with an Id, as I mentioned earlier). I don't need that here so will remove it but will need it later for get/delete API tests.

cbuescher replied:
Just also realize I haven't implemented get yet in this PR, we could change this later to a loop with get if you agree? 
I don't see the benefits over a long-enough wait time though, quite the oposite. I never got this call to _not_ finish in the first call, neither locally nor on CI which is even faster.

javanna replied:
I like the loop idea if it does not make things too complicated, it sounds doable.

javanna replied:
it's fine to make this a loop later ;)

javanna reviewed (commented):
I left a couple of comments but nothing major besides a question on exposing the inner search request. Thanks for picking this up!!!

Commit:
Adressing review comments

cbuescher commented:
@javanna thanks for the review, I pushed an update that should adress your comments, including hiding the wrapped inner SearchRequest. For that I unfortunately had to add a bunch of additional getters for all the affected parameters that need to be translated in the AsyncRequestConverter and tests. Let me know if you think its worth it.

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/SubmitAsyncSearchRequest.java:
```diff
@@ -0,0 +1,282 @@
+     */
+    public SearchSourceBuilder getSearchSource() {
+        return this.searchRequest.source();
+    }
```
we may want to expose a setter for the source too. Not so important but I guess users would argue that once set at the constructor it can never be reset. Search request allows to set it.

cbuescher replied:
Sorry, I don't follow. Why would you want to change the source when you are constructing the request in the client? You can get it and change the various parts of it if you like. Why aim for parity with SearchRequest? This would mean allowing the user to reach down into a thing we want to hide (the SearchRequest) and make changes there. I don't see how thats useful.

javanna replied:
I would have replied the same way to this comment :) it is just a comment that I got in the past with these client requests. People expect to be able to reset stuff, while here once you set the source you can not go back to the initial state. Though you can modify it. I am fine with doing without the setter though, it is not that useful.

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/SubmitAsyncSearchRequest.java:
```diff
@@ -0,0 +1,282 @@
+     */
+    public Boolean getRequestCache() {
+        return this.searchRequest.requestCache();
+    }
```
how about maxConcurrentShardRequests? Was it left out on purpose?

cbuescher replied:
Good one, I think that parameter is also missing from the regular RequestConverters.

cbuescher replied:
I pulled adding this parameter (and the previously mentioned pre_filter_shard_size) to the regular RequestConverters into https://github.com/elastic/elasticsearch/pull/53745 to get this in independently of this PR.

javanna replied:
sounds good thanks

cbuescher replied:
Cherry-picked that PR and added support for max_concurrent_shard_requests

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/SubmitAsyncSearchRequest.java:
```diff
@@ -0,0 +1,282 @@
+        }
+        if (searchRequest.isCcsMinimizeRoundtrips()) {
+            validationException.addValidationError("[ccs_minimize_roundtrips] is not supported on async search queries");
+        }
```
some of these checks become redundant once the search request is hidden. Are we going to be able to remove the getter for the search request, or will it be package protected? I guess that will play a role in whether these values are at least theoretically settable from the same package, or not at all.

cbuescher replied:
That getter can probably go, other need to be accessible from the AsyncRequestConverters, which currently need to stay in `org.elasticsearch.client` to work with existing infra in RequestConverters. Moving the former proved to be difficult. I think I can remove the validation checks for "ccs_minimize_roundtrips" and "scroll" though because now they should not be settable any more (and there is an extra layer of protection on the REST layer later.

cbuescher replied:
there should be no setters/getters for scroll and `ccs_minimize_roundtrips` anymore now so I removed the validation.

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/AsyncSearchRequestConverters.java:
```diff
@@ -56,4 +55,21 @@ static Request submitAsyncSearch(SubmitAsyncSearchRequest asyncSearchRequest) th
+        }
+        params.putParam("batched_reduce_size", Integer.toString(request.getBatchedReduceSize()));
+        params.putParam("pre_filter_shard_size", Integer.toString(SubmitAsyncSearchRequest.DEFAULT_PRE_FILTER_SHARD_SIZE));
+        params.putParam("ccs_minimize_roundtrips", Boolean.toString(SubmitAsyncSearchRequest.DEFAULT_CCS_MINIMIZE_ROUNDTRIPS));
```
should this be there given that we don't allow to set it directly and we support only one value for it?

cbuescher replied:
Yes, wee need to send those values (the onse we set on the ctor) via HTTP because they are different than the defaults in SearchRequest.

cbuescher replied:
Okay, upon rechecking the code in RestSearchAction#prepareRequest, we set those values in the ctor on the server side to the defaults and shold only send those that the user can still modify, like `batchedReduceSize` and `requestCache`(see https://github.com/elastic/elasticsearch/pull/53592#discussion_r393863122)

cbuescher replied:
Removing "pre_filter_shard_size" and "ccs_minimize_roundtrips" from the http serialization

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/AsyncSearchRequestConverters.java:
```diff
@@ -56,4 +55,21 @@ static Request submitAsyncSearch(SubmitAsyncSearchRequest asyncSearchRequest) th
+            params.putParam("allow_partial_search_results", Boolean.toString(request.getAllowPartialSearchResults()));
+        }
+        params.putParam("batched_reduce_size", Integer.toString(request.getBatchedReduceSize()));
+        params.putParam("pre_filter_shard_size", Integer.toString(SubmitAsyncSearchRequest.DEFAULT_PRE_FILTER_SHARD_SIZE));
```
I think we can remove it. it is not in the spec and we don't allow to set it directly

javanna reviewed (commented):
left a few more comments. I like it. Only a couple of nits around the changes made to hide the search request

Commit:
Remove redundant parameters and validation

Commit:
Add unsupported parameters to HLRC search request

Currently we don't send values for the `pre_filter_shard_size` and
`max_concurrent_shard_requests` SearchRequest parameters over http when using
the High Level Rest Client. This change adds these parameters to the
RequestConverters and tests.

Commit:
Avoid additional boilerplate getters on SubmitAsyncSearchRequest by making getSearchRequest() package private

Commit:
Revert "Avoid additional boilerplate getters on SubmitAsyncSearchRequest by making getSearchRequest() package private"

This reverts commit f686a5081f0a2489c32eae027c0b7817c18f0884.

Commit:
Add support for max_concurrent_shard_requests

Commit:
Merge branch 'master' into async-search-hlrc

cbuescher commented:
@elasticmachine run elasticsearch-ci/1

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/AsyncSearchClient.java:
```diff
@@ -0,0 +1,68 @@
+        this.restHighLevelClient = restHighLevelClient;
+    }
+
+    // TODO add docs url
```
docs link should be https://www.elastic.co/guide/en/elasticsearch/reference/current/async-search.html PR is open but not merged yet: #53675

cbuescher replied:
done, see ebd1906

javanna commented on client/rest-high-level/src/main/java/org/elasticsearch/client/AsyncSearchRequestConverters.java:
```diff
@@ -0,0 +1,74 @@
+        if (request.getAllowPartialSearchResults() != null) {
+            params.putParam("allow_partial_search_results", Boolean.toString(request.getAllowPartialSearchResults()));
+        }
+        params.putParam("batched_reduce_size", Integer.toString(request.getBatchedReduceSize()));
```
nit: maybe given they are now used in two places we could add the with* methods for these params too?

cbuescher replied:
done in f65cf82

javanna reviewed (approved):
LGTM

Commit:
Merge branch 'master' into async-search-hlrc

Commit:
Add docs urls

Commit:
Unify conversion of some request parameters

cbuescher commented:
Thanks @javanna, I pushed updates around your latest comments. @jimczi do you want to take another look or does this look okay now and I can start trying to get this through CI finally?

jimczi commented on client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java:
```diff
@@ -860,6 +860,26 @@ Params withPreference(String preference) {
+            return putParam("request_cache", Boolean.toString(requestCache));
+        }
+
+        Params withPartialResults(boolean allowPartialSearchResults) {
```
nit: rename to ` withAllowPartialResults` ?

jimczi commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/AsyncSearchResponse.java:
```diff
@@ -0,0 +1,224 @@
+        PARSER.declareBoolean(ConstructingObjectParser.constructorArg(), IS_RUNNING_FIELD);
+        PARSER.declareLong(ConstructingObjectParser.constructorArg(), START_TIME_FIELD);
+        PARSER.declareLong(ConstructingObjectParser.constructorArg(), EXPIRATION_FIELD);
+        PARSER.declareString(AsyncSearchResponse::setId, ID_FIELD);
```
Why can't this be a constructor arg ?

cbuescher replied:
True, this could be an `optionalConstructorArg()`. I think I did this via setters because I disklike too many optional ctor args that can be `null` and made the setters for them package private. I can even make them private I think, would that be okay with you? I don't remember exactly if more ctor args somehow slow down ConstructingObjectParser since they need to be buffered before object creation but I guess that would be a minor disadvantage.

cbuescher replied:
I pushed some changes and went with `optionalConstructorArg` to avoid more unnecessary roundtrips on this, in the end I don't care that much.

jimczi commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/AsyncSearchResponse.java:
```diff
@@ -0,0 +1,224 @@
+        PARSER.declareLong(ConstructingObjectParser.constructorArg(), EXPIRATION_FIELD);
+        PARSER.declareString(AsyncSearchResponse::setId, ID_FIELD);
+        PARSER.declareObject(AsyncSearchResponse::setSearchResponse, (p, c) -> AsyncSearchResponse.parseSearchResponse(p),
+                RESPONSE_FIELD);
```
Same here ?

jimczi commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/AsyncSearchResponse.java:
```diff
@@ -0,0 +1,224 @@
+        PARSER.declareString(AsyncSearchResponse::setId, ID_FIELD);
+        PARSER.declareObject(AsyncSearchResponse::setSearchResponse, (p, c) -> AsyncSearchResponse.parseSearchResponse(p),
+                RESPONSE_FIELD);
+        PARSER.declareObject(AsyncSearchResponse::setFailure, (p, c) -> ElasticsearchException.fromXContent(p), ERROR_FIELD);
```
And here ?

jimczi commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/AsyncSearchResponse.java:
```diff
@@ -0,0 +1,224 @@
+     * Sets the optional id property, use by parser and tests
+     */
+    void setId(String id) {
+        this.id = id;
```
This can be set in the ctr ? It should be final too.

jimczi commented on client/rest-high-level/src/main/java/org/elasticsearch/client/asyncsearch/AsyncSearchResponse.java:
```diff
@@ -0,0 +1,224 @@
+     * Sets the optional search response property, used by parser and tests
+     */
+    void setSearchResponse(SearchResponse searchResponse) {
+        this.searchResponse = searchResponse;
```
same here

jimczi reviewed (approved):
I left some nit comments regarding the async search response parsing but the code looks good to me.

Commit:
adress review comments

jimczi reviewed (approved):
LGTM, thanks @cbuescher !

cbuescher commented:
@elasticmachine run elasticsearch-ci/1

Commit:
Merge branch 'master' into async-search-hlrc

cbuescher commented:
@elasticmachine run elasticsearch-ci/bwc

cbuescher commented:
@elasticmachine run elasticsearch-ci/default-distro

cbuescher commented:
@elasticmachine run elasticsearch-ci/1
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/default-distro

cbuescher commented:
@elasticmachine run elasticsearch-ci/packaging-sample-matrix-unix
@elasticmachine run elasticsearch-ci/1

Commit:
iter
